### ISSUE TITLE
Snapping randomly not working

### ISSUE THREAD

**Issue by dstoyanoff** [rxhanson/Rectangle#1270]
- [x] This is a new bug that hasn't been documented in a closed issue or in the Readme.

macOS version: Sonoma
Rectangle version: 0.72 (78)
Logs if applicable (In Rectangle menu, hold option, "View Logging..."): no logs

I saw that you released a version yesterday that was supposed to fix snapping, but I am still having issues with it and it only works from time to time. Can't isolate when it stops working

**Comment by Danny-mtb** [comment]
I have the same problem after the last update, it only works shortly after a restart and then no longer works

MacOS: Sonoma 14.0
Rectangle Pro version: 3.0.11 (152)

**Comment by dstoyanoff** [comment]
@rxhanson, the thing is that it is random and not associated to a particular app. Happens with basic apps like the mail client. Then all of a sudden it starts working without even me restarting the app. I'll keep trying to understand what happens and let you know if I discover something more.

**Comment by dstoyanoff** [comment]
Interesting thing is for me restarting the app doesn't fix the issue. I have left the logging ongoing until the issue occur, here is the result:

```
2023-09-26T10:45:54+03:00: Unable to obtain the accessibility element with the specified attribute at mouse location
2023-09-26T10:45:54+03:00: Unable to obtain the accessibility element with the specified attribute at mouse location
2023-09-26T10:45:56+03:00: Window under cursor fallback matched: Safari
2023-09-26T10:45:56+03:00: Window under cursor fallback matched: Safari
2023-09-26T10:45:56+03:00: Failed to find frontmost window.
2023-09-26T10:45:58+03:00: Unable to obtain window info from location
2023-09-26T10:45:58+03:00: Window under cursor fallback matched: Google Chrome
2023-09-26T10:45:59+03:00: AX sizing proposed: (1706.6666666666667, 1360.0), result: (1707.0, 1360.0)
2023-09-26T10:45:59+03:00: AX position proposed: (1706.0, 25.0), result: (1706.0, 25.0)
2023-09-26T10:45:59+03:00: AX sizing proposed: (1706.6666666666667, 1360.0), result: (1707.0, 1360.0)
2023-09-26T10:45:59+03:00: centerThird | display: (0.0, 55.0, 5120.0, 1360.0), calculatedRect: (1706.0, 25.0, 1706.6666666666667, 1360.0), resultRect: (1706.0, 25.0, 1707.0, 1360.0), srcScreen: C49RG9x, destScreen: C49RG9x, resultScreen: C49RG9x
2023-09-26T10:48:52+03:00: Unable to obtain window info from location
2023-09-26T10:48:52+03:00: Unable to obtain the accessibility element with the specified attribute at mouse location
2023-09-26T10:49:39+03:00: Unable to obtain window info from location
2023-09-26T10:49:39+03:00: Unable to obtain the accessibility element with the specified attribute at mouse location
2023-09-26T10:49:39+03:00: Unable to obtain window info from location
2023-09-26T10:49:39+03:00: Unable to obtain the accessibility element with the specified attribute at mouse location
2023-09-26T10:55:31+03:00: Unable to obtain the accessibility element with the specified attribute at mouse location
2023-09-26T10:55:31+03:00: Unable to obtain the accessibility element with the specified attribute at mouse location
2023-09-26T10:55:34+03:00: Window under cursor fallback matched: Slack
2023-09-26T11:08:30+03:00: Unable to obtain the accessibility element with the specified attribute at mouse location
```

I don't really see anything different to be honest.

`Unable to obtain the accessibility element with the specified attribute at mouse location` is shown on mouse click on the menu bar (maybe in some other cases too)

Rest of the errors are also observed before and after snapping stopped working, so it is either not related or the error is the same in multiple occasions

**Comment by dstoyanoff** [comment]
I've been able to actually isolate one of the occasions of the issue and it is with notifications. I use reminders a lot and while there is a visible reminder, snapping doesn't work. Closing the reminder fixes it. 

Will monitor for other causes as I think this is not the only one.

**Comment by rxhanson** [comment]
@dstoyanoff Would you mind using this build? (you'll have to bypass gatekeeper to install it since the auto-builds aren't signed. There's no harm in using it, or you can wait until I create a release) 

https://github.com/rxhanson/Rectangle/actions/runs/6312463594

That build will ignore Notification Center and has an additional log statement that will tell us which window is being found under the cursor.

**Comment by dstoyanoff** [comment]
Unfortunately it didn't work. Here is the log:

```
2023-09-26T15:36:07+03:00: Window found at (4323.9140625, 53.328125): WindowInfo(id: 20, level: 23, frame: (0.0, 0.0, 5120.0, 1440.0), pid: 822, processName: Optional("Notification Centre"))
```

**Comment by dstoyanoff** [comment]
Seems it should be `Notification Centre`, while you ignored `Notification Center`

**Comment by rxhanson** [comment]
Thanks for the quick test! Here's another one, where rather than keying off process name (obviously fragile), I'm lowering the acceptable window level below that of Notification Center.

https://github.com/rxhanson/Rectangle/actions/runs/6313160287

**Comment by dstoyanoff** [comment]
Thank you for the fast delivery. This build seems to be working. I will be monitoring if it stops working in some other occasion. Thanks for the quick reaction!

**Comment by Zackhardtoname** [comment]
In my case, snapping doesn't work for my secondary monitor (of a smaller size than the primary one).
All operations would happen at the primary one instead. It just randomly started these last few days and I have not updated Rectangle or Mac explicitly.

**Comment by rxhanson** [comment]
@Zackhardtoname, which version of Rectangle are you using?

**Comment by Zackhardtoname** [comment]
> @Zackhardtoname, which version of Rectangle are you using?

About that... I thought an update might fix it so I just updated Rectangle to the latest without noting the version...

**Comment by rxhanson** [comment]
@Zackhardtoname I went ahead and released the version discussed in this issue as v0.73, so you can try that one and see if it fixes the issue you're having.

**Comment by Zackhardtoname** [comment]
It didn't. Mac OS version is 13.5 btw.

**Comment by Zackhardtoname** [comment]
Maybe it was after reconnecting the cables, mine started working smoothly again (without any machine restarts)

**Comment by rxhanson** [comment]
@Zackhardtoname interesting, thanks for following up. I suspect this would be an issue in what macOS is communicating to Rectangle, and might not be something fixable from within Rectangle. Let me know if you hit the issue again - maybe there's something more we can figure out.

**Comment by XadRav** [comment]
This is happening to me too.  I usually use my MacBook Air with a thunderbolt dock and second monitor.  If I try to snap a window that is already on the second monitor, then it's 50/50 whether snapping works.  But if I drag and drop the window to my laptop screen, then move it back to the second monitor it seems to snap every time.

Version: 0.73
macOS Sonoma 14.1
Log is empty when snapping fails.

**Comment by vani2** [comment]
Same for me, tried to reinstall.
Version: 0.74 (80)
macOS Sonoma 14.1
Log is empty when snapping fails too

**Comment by XadRav** [comment]
Yep just updated to version 0.74 and same thing is happening.

**Comment by rxhanson** [comment]
If you’ve recently updated macOS, you’ll want to reboot your Mac one more time.

**Comment by touvex** [comment]
First of all, absolutely love this app, thank you for your work! I honestly don't know if i can even use a mac in my workflow without this...

This issue have just started to happen to me too, also randomly.

I'm using:

Sonoma 14.1.1
Rectangle 0.74 (80)

orginally i thought it was because i was using an old OS Vnetura, so i updated today but it did not fix the issue.

Thanks for your help.

**Comment by rxhanson** [comment]
@touvex, do you see any logs for this? You can hold the option key in the menu bar menu and select view logging, then try to reproduce the issue.

**Comment by touvex** [comment]
> @touvex, do you see any logs for this? You can hold the option key in the menu bar menu and select view logging, then try to reproduce the issue.

Thank you, I think somehow, it has fixed itself, fingers crossed, maybe i just needed that restart after updating OS.

I think it may have been Microsoft Teams NEW, i switched back to Teams Classic and it has not had issues since, i will keep monitoring and post here if that changes! Thanks again, your app is amazing!

**Comment by touvex** [comment]
PS. i actually didnt know you had pro, as i went to donate, so have now bought the pro, do you advise i uninstall normal first and then install pro?

**Comment by tibeer** [comment]
Can confirm that Microsoft Teams NEW has definetly an impact. Though it seems to be inconsistent across colleagues.
Problems started popping up once they installed the "NEW" version of teams, but not for all of them. I guess this is not a rectangle problem then but rather Teams doing suspicious things .... :/

**Comment by rxhanson** [comment]
@touvex Thanks for purchasing pro! Sorry I never responded, but for future reference I just recommend not having them running at the same time.

@tibeer, thanks for reporting this. There have been some adjustments to Rectangle in recent releases, and you can try this terminal command to revert to an old way of how Rectangle would obtain the window under the cursor:
`defaults write com.knollsoft.Rectangle systemWideMouseDown -int 1`

In short, the old way can break click event receiving for certain apps for some users, but the new way can make it so snapping doesn't work for some apps for certain users. At least in the new way, it's not as frustrating for the small percentage of users that run into an issue. It appears that most users are fine with either way is used.

**Comment by tibeer** [comment]
Perfect! Will try it out once myself or colleagues run into this. By the way I forgot to mention that keyboard shortcuts work all the time, so it might be pretty much the mouse thing :)

**Comment by ra-beer** [comment]
Hey there. Here is my solution what works for me: You must pick the window very left near the corner and move it where you want like before. You can't pick everywhere you want like before. But this is a solution for me that help me during my daily business. It's a little kind of spooky, but it is a workaround :)

**Comment by XadRav** [comment]
@rxhanson Just to add another data point, I have confirmed that it is "New" Teams causing this issue for me as well.  This is very repeatable with the bug consistently happening while New Teams is running, and goes away when I quit New Teams or while running Old Teams.  The terminal command you posted also seems to fix the issue with new Teams running.  Would it be possible in future releases to perhaps add an option for that command in preferences so I won't need to find this post to find the command in the future?  Thanks for all your hard work!

**Comment by rxhanson** [comment]
@XadRav, I don't have plans to add this into the UI, mainly because it would be really hard to convey concisely enough to add it in. However, if I add a "hidden settings" sort of UI at some point, then I'll put it there.

**Comment by Victorious3** [comment]
> Can confirm that Microsoft Teams NEW has definetly an impact. Though it seems to be inconsistent across colleagues. Problems started popping up once they installed the "NEW" version of teams, but not for all of them. I guess this is not a rectangle problem then but rather Teams doing suspicious things .... :/

I have the same issue and I can confirm that it goes away when quitting teams. Very strange.

**Comment by HackettLai** [comment]
I'm having the same issue on the latest version as well. It seems to happen only on my secondary monitor. I tried to snap Finder to the edge, but nothing happened. Here is my screenshot of the log for investigation.
![CleanShot 2024-06-24 at 16 45 11](https://github.com/rxhanson/Rectangle/assets/56993446/b54e3871-4a8b-4cc5-bf78-afd0042345fd)

The last successful attempt was on the built-in display.
- App version: v0.80 (86)
- macOS 14.5 (23F79)

**Comment by rxhanson** [comment]
@HackettLai Are the initial logs the ones where it's not working, where it's unable to obtain the accessibility element? It is not common for this to happen on Finder, do you see this with all apps, and is there anything unusual about your setup?

**Comment by HackettLai** [comment]
> @HackettLai Are the initial logs the ones where it's not working, where it's unable to obtain the accessibility element? It is not common for this to happen on Finder, do you see this with all apps, and is there anything unusual about your setup?

Hi @rxhanson I've played around with it for a day and even reinstalled macOS. It seems like I can now provide a more detailed description of the issue. I've recorded a video for better investigation: [YouTube Link](https://youtu.be/dfEe9nQGHN0?si=hbEEK7O2XgJugOlU.)

**Symptom**

1. While recording the video, it seems like the issue happens on my built-in display.
2. I can snap to position on both sides of my display, but the snap area cannot be recognized on the built-in display.
3. Sometimes it can be recognized once but fails on the recognition again afterward.
4. I've tried both Finder and Notion (at 1:17), and both applications produce the same result.

**App Setting**
![CleanShot 2024-06-26 at 10 01 04](https://github.com/rxhanson/Rectangle/assets/56993446/ecf32ca8-1ff0-4cf4-a3d5-414a874de746)
![CleanShot 2024-06-26 at 10 01 07](https://github.com/rxhanson/Rectangle/assets/56993446/111a54e4-0aaf-49d8-b6a4-83ec1a563089)

**Information**
- App version: v0.80 (86)
- macOS 14.5 (23F79)

**Comment by rxhanson** [comment]
@XylotoGhost The recommendation is indeed still the same, execute this Terminal command and restart the app:

`defaults write com.knollsoft.Rectangle systemWideMouseDown -int 1`

Since you brought this one back up, I'm thinking it might make sense to automatically apply this behavior to MS Teams.

**Comment by rxhanson** [comment]
Currently this behavior is actually already applied to MS Teams automatically. At this point, if you're hitting issues with MS Teams, then this is something best addressed by Microsoft. Closing this one out, but feel free to open a new issue for further discussion.

**Comment by RickyRicardo17** [comment]
Noting that this issue is still present but is fixed by closing any notification popups (in my case a chrome notification).

**Comment by XylotoGhost** [comment]
I agree that the issue has not been fixed. For me it occurs also without an open notification. Microsoft Teams is the only app where I notice this behaviour.

**Comment by larsgerber** [comment]
I observed the following behavior regarding window snapping with Microsoft Teams.

For each of the actions below, I verified whether snapping still works afterward:

- Minimizing via the yellow button (top left corner) → works
- Minimizing via `CMD + M` → works
- Hiding via `CMD + H `→ works
- Soft closing the window via `CMD + W` → does not work
- Soft closing the window via the red button (top left corner) → does not work

The behavior is the same regardless of whether there are open notifications in the Notification Center.

I restarted Microsoft Teams after each test to ensure consistent results. Restarting Rectangle is also possible.

**Comment by rxhanson** [comment]
@XylotoGhost does window snapping work in Microsoft Teams, or work everywhere except Microsoft Teams?

If it's working in Teams but nowhere else, then this command typically solves the issue:
`defaults write com.knollsoft.Rectangle systemWideMouseDown -int 1`
If it's working everywhere except Teams, then you can set it to 2:
`defaults write com.knollsoft.Rectangle systemWideMouseDown -int 2`

Restart the app after.

Do you still have issues after executing that and restarting?


@larsgerber, thanks for reporting. Can you also try this Terminal command setting and restarting the app?
`defaults write com.knollsoft.Rectangle systemWideMouseDown -int 2`

**Comment by larsgerber** [comment]
Thank you @rxhanson.
It works like explained in https://github.com/rxhanson/Rectangle/issues/1692#issuecomment-3758293198. Restarting both apps already fixed the issue.
